<a href="https://colab.research.google.com/github/zelal-Eizaldeen/data_qaulity_course/blob/main/6_duplication.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

Create a sample dataset with employee information. We’ll intentionally introduce duplicate employee IDs to demonstrate the identification of duplicates:

In [3]:
data = {
    'EmployeeID': [101, 102, 103, 101, 104, 105, 102],
    'FirstName': ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Frank', 'Bob'],
    'LastName': ['Smith', 'Johnson', 'Brown', 'Davis', 'Lee', 'White', 'Johnson'],
}
df = pd.DataFrame(data)

We’ll use pandas to identify and mark duplicate records based on the EmployeeID column. The duplicated() function is used to create a Boolean mask, where True indicates a duplicated record:

In [10]:
duplicated_mask = df.duplicated(subset='EmployeeID', keep='first')
duplicated_mask

0     True
1     True
2    False
3     True
4    False
5    False
6     True
dtype: bool

The subset='EmployeeID' argument specifies the column on which the duplication check is performed. keep='first' marks duplicates as True except for the first occurrence. You can change this parameter to last or False based on your requirements.

We then create a new column called 'IsDuplicate' in the DataFrame to indicate whether each record is a duplicate or not:

In [5]:
df['IsDuplicate'] = duplicated_mask

We calculate the percentage of duplicate records by dividing the number of duplicate records (those marked as True in the 'IsDuplicate' column) by the total number of records and then multiplying by 100 to express it as a percentage:

In [11]:
duplicate_percentage = (df['IsDuplicate'].sum() / len(df)) * 100


we display the dataset with the IsDuplicate column to see which records are duplicates. Here’s the final output:

In [12]:
print(df)
print(f"Percentage of Duplicate Records: {duplicate_percentage}")

   EmployeeID FirstName LastName  IsDuplicate
0         101     Alice    Smith        False
1         102       Bob  Johnson        False
2         103   Charlie    Brown        False
3         101     David    Davis         True
4         104       Eve      Lee        False
5         105     Frank    White        False
6         102       Bob  Johnson         True
Percentage of Duplicate Records: 28.57142857142857


This output indicates that 28.57% of records are duplicated in the dataset.

- The fewer duplicated records, the better!
- The threshold for what is considered an acceptable or “good” level of duplicated records in a dataset can vary depending on the specific context and the goals of your data management or analysis. There is no one-size-fits-all answer to this question, as it depends on factors such as the type of data, the purpose of the dataset, and industry standards.